src/charm.py

```diff
@@ -9,93 +9,306 @@
+
+        if self.is_kafka_related:
+            result[KAFKA] = list(self.kafka.fetch_relation_data().values())[0]
+
         event.set_results(result)
 
     def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
         """Event triggered when a database was created for this application."""
-        logger.info(f"database credentials are received: {event.username}")
-        self.unit.status = ActiveStatus(f"received {self.database} credentials")
+        logger.debug(f"database credentials are received: {event.username}")
+        self._on_config_changed(event)
+        if not self.unit.is_leader:
+            return
+        # update values in the databag
+        self._update_active_relations(event, "active")
+
+    def _on_topic_created(self, event: TopicCreatedEvent) -> None:
+        """Event triggered when a topic was created for this application."""
+        logger.debug(f"Kafka credentials are received: {event.username}")
+        self._on_config_changed(event)
+        if not self.unit.is_leader:
+            return
+        # update status of the relations in the peer-databag
+        self._update_active_relations(event, "active")
+
+    def _update_active_relations(self, event: RelationEvent, status: STATUSES) -> None:
+        """Update the relation status in the peer-relation databag."""
+        self.set_secret("app", f"relation-{ event.relation.id}", status)
+
+    def _on_peer_relation_changed(self, event: RelationEvent) -> None:
+        """Handle the peer relation changed event."""
+        if not self.unit.is_leader:
+            return
+        removed_relations = []
+        # check for relation that has been removed
+        for relation_data_key, relation_value in self.app_peer_data.items():
+            if "relation-" in relation_data_key:
+                if relation_value == "broken":
+                    removed_relations.append(relation_data_key)
+
+        if removed_relations:
+            # update configuration
+            self._on_config_changed(event)
+            # update relation status to removed
+            for relation in removed_relations:
+                self.set_secret("app", relation, "removed")
+
+    @property
+    def topic_name(self) -> Optional[str]:
+        """Return the configured topic name."""
+        return self.model.config.get("topic-name", None)
+
+    @property
+    def database_name(self) -> Optional[str]:
+        """Return the configured database name."""
+        return self.model.config.get("database-name", None)
+
+    @property
+    def extra_user_roles(self) -> Optional[str]:
+        """Return the configured extra user roles."""
+        return self.model.config.get("extra-user-roles", None)
+
+    @property
+    def kafka_relation(self) -> Optional[Relation]:
+        """Return the kafka relation if present."""
+        return self.kafka.relations[0] if len(self.kafka.relations) > 0 else None
```

Comment:
**nitpick:** This can be `return self.kafka.relations[0] if len(self.kafka.relations) else None`. 0 is falsie.